lib/graphql/execution/interpreter/runtime.rb

```diff
@@ -333,6 +345,9 @@ def set_result(selection_result, result_name, value)
                 @response = nil
               else
                 set_result(parent, name_in_parent, nil)
+                # This is odd, but it's how it used to work. Even if `parent` _would_ accept
+                # a `nil`, it's marked dead. TODO: check the spec, is there a reason for this?
+                parent.graphql_dead = true
```

Comment:
Yeah this is weird to me too, I would expect it to just be `selection_result.graphql_dead = true` since it's the current field that actually needs marking as dead?

But I note there's something going on I don't grok with `selection_result.graphql_non_null` where it's either a boolean (whether the current field is non-null or not, makes sense, though maybe it ought to be a reference to the query AST type?) _or_ it can be a list of something for reasons that aren't clear to me. Maybe we're not always operating directly on `selection_result` as a single field or something?

Reply:
Also what is the difference between this case in `set_result` and in `continue_value`, they both seem to handle part of the invalid-nil flow?

Reply:
Regarding `graphql_non_null`, it was basically used in two distinct modes: `true` indicated that the selection_result object was an Array which allowed no `nil`s; `Array<String>` indicated that the selection_result was a Hash which didn't allow `nil`s for the fields whose names were contained in the Array. 

It was straightforward to highlight this distinction by using two separate accessors, so I did so in a322abaac

Reply:
Btw, interesting consideration which I haven't explored: `graphql_non_null_fields` is currently an array, but for queries that include aliases, that array may contain duplicates. It might be that a Set is better for this. In practice I think it depends on size, so I'm not sure the best way to approach it.

Reply:
I take that back -- it uses `result_name`, which would be the alias if there is one. So it wouldn't ever contain duplicates. However, whenever the same selection (from the AST) is executed, it will build up duplicate arrays 🤔